src/draw/lv_img_src.c

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file lv_img.c
+ *
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_img_src.h"
+#include "lv_img_buf.h"
+
+#include "../misc/lv_assert.h"
+
+static lv_res_t alloc_str_src(lv_img_src_t * src, const char * str);
+
+static const char * find_ext(const char * src);
+
+/**********************
+ * GLOBAL PROTOTYPES
+ **********************/
+
+/**
+ * Get the type of an image source
+ * @param src pointer to an image source:
+ *  - pointer to an 'lv_img_t' variable (image stored internally and compiled into the code)
+ *  - a path to a file (e.g. "S:/folder/image.bin")
+ *  - or a symbol (e.g. LV_SYMBOL_CLOSE)
+ * @return type of the image source LV_IMG_SRC_VARIABLE/FILE/SYMBOL/UNKNOWN
+ * @deprecated You shouldn't rely on this function to find out the image type
+ */
+lv_img_src_type_t lv_img_src_get_type(const void * src)
+{
+    lv_img_src_type_t img_src_type = LV_IMG_SRC_UNKNOWN;
+
+    if(src == NULL) return img_src_type;
+    const uint8_t * u8_p = src;
+
+    /*The first byte shows the type of the image source*/
+    if(u8_p[0] >= 0x20 && u8_p[0] <= 0x7F) {
+        img_src_type = LV_IMG_SRC_FILE; /*If it's an ASCII character then it's file name*/
+    }
+#if LVGL_IMG_ALLOW_TYPELESS_SRC
+    else if(u8_p[0] == 0xFF) {
+        img_src_type = LV_IMG_SRC_OBJ;
+    }
+#endif
+    else if(u8_p[0] >= 0x80) {
+        img_src_type = LV_IMG_SRC_SYMBOL; /*Symbols begins after 0x7F*/
+    }
+    else {
+        img_src_type = LV_IMG_SRC_VARIABLE; /*`lv_img_dsc_t` is draw to the first byte < 0x20*/
+    }
+
+    if(LV_IMG_SRC_UNKNOWN == img_src_type) {
+        LV_LOG_WARN("lv_img_src_get_type: unknown image type");
+    }
+
+    return img_src_type;
+}
+
+lv_res_t lv_img_src_parse(lv_img_src_t * uri, const void * src)
```

Comment:
Shouldn't it have similar prototype to lv_img_src_set_symbol?
E.g. 

```suggestion
lv_res_t lv_img_src_set_dsc(lv_img_src_t * src, const void * src_data)
```

TBH, I think `uri` is not descriptive enough for novice users.

Reply:
I think there a misunderstanding here. `lv_img_src_parse` is the function used to accept the *old* void* parameter that could point to either a file, a dsc or a symbol. This function is used for compatibility where void* are kept (like bg_img_src, window_icon, etc...). Ideally, it should disappear in LVGL9, but that's a lot of work since it implies modifying the API to remove all traces of void* for image source (and instead replace by `lv_img_src_t*`).

Again we have an issue here because since everything converts to void* a code written for LVGL <= 8 will compile in LVGL9 if we keep the function name the same, but crash at runtime.

So we need to agree on how to rename the functions in order to avoid this issue. I'm not that good at function name, if you have some ideas please suggest!

Reply:
I've tried different strategy to keep the same name but have the compiler to break on the new code:
Ideally we would want to go from `lv_img_set_src(lv_img_t * img, void* src)` to `lv_img_set_src(lv_img_t * img, lv_img_src_t * src)` but since `void*` is (silently) convertible to the `lv_img_src_t*` it'll silently compile invalid code like `lv_img_set_src(img, dsc->bg_img)` where dsc->bg_img is a void * containing a symbol or a file's path.

Here are the strategy I've thought about:
1. Change the type only and add a `-Wc++-compat` flag to the compilation. This will issue a compiler warning on GCC (sorry for other compiler). The code will still build and crash at runtime.
2. Change the signature to add another parameter (like `lv_img_set_src(lv_img_t * img, lv_img_src_t * src, uint32_t flags)`. In that case, users will have to adapt their code (so a giant warning above the function declaration should say "don't use void * anymore") 
3. Change the function name (like `lv_img_set_src_ex(lv_img_t * img,  lv_img_src_t * src)` and remove the previous function. 
4. Add a new function name and keep the previous function that'll go to a backport parsing (like in this PR). However, user will continue to use the old function until it's removed. And this might not work for every place a `void *` is used to store some *guess what I am* source. 
5. Pass src type by value, like (`lv_img_set_src(lv_img_t * img, lv_img_src_t src)`). That's probably the best solution but not as efficient as passing a pointer. Since the `lv_img_src_t` is a small structure (one type, one pointer and a length), that might work ok anyway. The compiler will refuse to compile previous code with a void * and so it should be easy to deal with. Also, we can have functions like this: `lv_img_src_t lv_img_src_from_file(const char* path)` to the final code would look like `lv_img_set_src(img, lv_img_src_from_file("/path/to/img.png"))` which I like because it's self describing.

What do you think is best ?

Reply:
> Pass src type by value, like (lv_img_set_src(lv_img_t * img, lv_img_src_t src)). That's probably the best solution but not as efficient as passing a pointer. Since the lv_img_src_t is a small structure (one type, one pointer and a length), that might work ok anyway.

I also like it the most. The current image converter can generate an `lv_img_src_t` variable too, so in this case:
```c
lv_img_set_src(img, my_img);  //not &my_img
```

It important to keep source type open for extension by external modules. E.g. the SDL engine might provide an `SDL_Texture` image source.

Reply:
Ok. I agree.

Right now, in LVGL9, lv_img_src_t is like this:
```cpp
struct src
{
    uint32_t type: 2;
    uint32_t length: 30; (or 62 on 64 bits machine)
    void * data;
};
```
So we have 4 possibles source type (file, data, symbol, unknown) and 1GB of for the source's data length (for data mainly). I think it's completely safe to reduce to only 512MB here and allow 8 possible source type. Storing a 512MB binary picture inside the final executable would be crazy anyway. 
That way, we could add new source type like "widget snapshot" and GPU textures.

Reply:
I'd rather choose:
```c
struct src
{
    uint32_t type: 8;
    uint32_t reserved: 24;
    uint32_t lenghth;
    void * data;
};
```

My way of thinking is that we are just redesigning an interface an seemingly we are close to some possible limits:
- having space only for 4 extra source types?
- only 512MB files? It it really that much? Even a 720p movie is larger than that.

Reply:
Length is for embedded data (in the final binary software). File & stream will store the path/uri length here, it's not the file size. So having 512MB of embedded resource in a binary .EXE is huge (IMHO). Even 256MB would be huge in today's standard.

I'm trying to limit the size of the structure since it'll be passed by value, so I think 2 or 3 registers is ok, but having more is not. There's no point in reserving space here, IMHO.  We could also steal bits from the data pointer, since it's always aligned, we can steal the 2 or 3 low bits.

How many type do you think of ?

1. uri (file path at first, can become true url in the future since ffmpeg supports them)
2. embedded data 
3. text (symbols)
4. snapshot (ptr to lv_img_dsc)
5. unknown
6. file descriptor (like GPU textures, DMABUF etc...)
7. ???
8. ???

Reply:
> File & stream will store the path/uri length here, 

Okay, I'm convinced :slightly_smiling_face: 


> How many type do you think of ?

I found some candidates. (I don't say all these makes sense and should be differentiated at `src_type` level but it shows that there are other options)

5. Vector path
6. SDL textures
6. OpenGL textures
6. DirectX textures
7. Some kind of stream?
8. Web URL
9. FTP
10. 3D model
10. 3D model in other format

I think web URL's should have it's own type e.g. to allow the PNG decoder to drop them. 

So 16 types seems better to me. >256 MB of embedded date really doesn't make much sense.
```c
struct src
{
    uint32_t type: 4;
    uint32_t length: 28; //256MB 
    void * data;
};
```

Anyway, this `struct` can be modified/adjusted at any time without breaking anything.

Reply:
Agreed. Will be in changed in LVGL9 with the change in the `lv_img_set_src` api. Right now, I've removed all version checking so that there isn't any API breakable at this level.